what4/src/What4/Protocol/SMTLib2.hs

```diff
@@ -884,6 +919,339 @@ parseBvArraySolverValue w v (SApp ["store", arr, idx, val]) = do
+  deriving (Functor, Applicative, Monad, MonadIO, MonadError String, MonadReader (ProcessorEnv sym))
+
+data ProcessorEnv sym = ProcessorEnv
+  { procSym :: sym
+  , procLetEnv :: HashMap Text (Some (I.SymExpr sym))
+  }
+
+runProcessor :: ProcessorEnv sym -> Processor sym a -> IO (Either String a)
+runProcessor env (Processor action) = runReaderT (runExceptT action) env
+
+opTable :: I.IsSymExprBuilder sym => HashMap Text (Op sym)
+opTable = HashMap.fromList
+  -- Boolean ops
+  [ ("not", Op1 knownRepr I.notPred)
+  , ("=>", Op2 knownRepr (Just RightAssoc) I.impliesPred)
+  , ("and", Op2 knownRepr (Just LeftAssoc) I.andPred)
+  , ("or", Op2 knownRepr (Just LeftAssoc) I.orPred)
+  , ("xor", Op2 knownRepr (Just LeftAssoc) I.xorPred)
+  -- Integer ops
+  , ("-", Op2 knownRepr (Just LeftAssoc) I.intSub)
+  , ("+", Op2 knownRepr (Just LeftAssoc) I.intAdd)
+  , ("*", Op2 knownRepr (Just LeftAssoc) I.intMul)
+  , ("div", Op2 knownRepr (Just LeftAssoc) I.intDiv)
+  , ("mod", Op2 knownRepr Nothing I.intMod)
+  , ("abs", Op1 knownRepr I.intAbs)
+  , ("<=", Op2 knownRepr Nothing I.intLe)
+  , ("<", Op2 knownRepr Nothing I.intLt)
+  , (">=", Op2 knownRepr Nothing $ \sym arg1 arg2 -> I.intLe sym arg2 arg1)
+  , (">", Op2 knownRepr Nothing $ \sym arg1 arg2 -> I.intLt sym arg2 arg1)
+  -- Bitvector ops
+  , ("bvnot", BVOp1 I.bvNotBits)
+  , ("bvneg", BVOp1 I.bvNeg)
+  , ("bvand", BVOp2 (Just LeftAssoc) I.bvAndBits)
+  , ("bvor", BVOp2 (Just LeftAssoc) I.bvOrBits)
+  , ("bvxor", BVOp2 (Just LeftAssoc) I.bvXorBits)
+  , ("bvadd", BVOp2 (Just LeftAssoc) I.bvAdd)
+  , ("bvsub", BVOp2 (Just LeftAssoc) I.bvSub)
+  , ("bvmul", BVOp2 (Just LeftAssoc) I.bvMul)
+  , ("bvudiv", BVOp2 Nothing I.bvUdiv)
+  , ("bvurem", BVOp2 Nothing I.bvUrem)
+  , ("bvshl", BVOp2 Nothing I.bvShl)
+  , ("bvlshr", BVOp2 Nothing I.bvLshr)
+  , ("bvsdiv", BVOp2 Nothing I.bvSdiv)
+  , ("bvsrem", BVOp2 Nothing I.bvSrem)
+  , ("bvashr", BVOp2 Nothing I.bvAshr)
+  , ("bvult", BVComp2 I.bvUlt)
+  , ("bvule", BVComp2 I.bvUle)
+  , ("bvugt", BVComp2 I.bvUgt)
+  , ("bvuge", BVComp2 I.bvUge)
+  , ("bvslt", BVComp2 I.bvSlt)
+  , ("bvsle", BVComp2 I.bvSle)
+  , ("bvsgt", BVComp2 I.bvSgt)
+  , ("bvsge", BVComp2 I.bvSge)
+  ]
+
+parseExpr ::
+  forall sym . I.IsSymExprBuilder sym => sym -> SExp -> Processor sym (Some (I.SymExpr sym))
+parseExpr sym sexp = case sexp of
+  "true" -> return $ Some $ I.truePred sym
+  "false" -> return $ Some $ I.falsePred sym
+  _ | Just i <- parseIntSolverValue sexp ->
+      liftIO $ Some <$> I.intLit sym i
+    | Just (Pair w bv) <- parseBVLitHelper sexp
+    , Just LeqProof <- testLeq (knownNat @1) w ->
+      liftIO $ Some <$> I.bvLit sym w bv
+  SAtom nm -> do
+    env <- asks procLetEnv
+    case HashMap.lookup nm env of
+      Just expr -> return $ expr
+      Nothing -> throwError ""
+  SApp ["let", SApp bindings_sexp, body_sexp] -> do
+    let_env <- HashMap.fromList <$> mapM
+      (\(SApp [SAtom nm, expr_sexp]) -> do
```

Comment:
Nit pick: this will trigger `-Wincomplete-uni-patterns` on GHC 9.2. We should provide a fall-through case as well to avoid warnings.